053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,457 @@
+
+  ```go
+  builds, _ := pipeline.Builds()
+  eventStore.Delete(builds)
+  ```
+
+  ...this doesn't work so well for the Postgres implementation. Currently, when
+  we delete a Pipeline from the database, `TRIGGER`s will `DROP` the
+  `pipeline_build_events_x` table for the appropriate pipeline. This results in
+  much more performant build event deletions, since dropping a table is faster
+  than deleting many thousands/millions of rows. I'd be happy to hear any clever
+  ideas anyone has on how to just have a single `Delete` method that acts on
+  builds while achieving the same table-dropping ability for Postgres.
+
+* `DeleteTeam(team db.Team)` deletes all builds for a given team. Same
+  reasoning as `DeletePipeline()`.
+
+* `UnmarshalKey(data []byte, key *Key)` works like `UnmarshalJSON` - give it
+  some bytes and tell it where you want to put the result. This is used e.g.
+  for the `/api/v1/builds/build_id}/events` endpoint where you can pass in a
+  `Last-Event-ID` header - we would call `UnmarshalKey` on this raw value, and
+  the `EventStore` would parse it appropriately. For instance, for Postgres, we
+  would expect `data` to look like a number in ASCII, so we could parse it
+  appropriately.
+
+  This is only necessary if we care about the `Last-Event-ID` header from what
+  I can tell (it doesn't appear that any internal code requires starting events
+  from a certain point - hence no need for a `MarshalKey` method). I left as an
+  open question whether we want to keep `Last-Event-ID`.
+
+
+## <a name="event-store-composition">`EventStore` Composition</a>
+
+In addition to enabling Concourse to interact with external stores besides
+Postgres, we can construct `EventStore` implementations that compose other
+`EventStores` to do some interesting things. One application is data
+transformations. I'll take the example of secret redaction:
+
+```go
+type SecretRedactingEventStore struct {
+    EventStore
+    // Some cred vars tracking state
+    ...
+}
+
+func (s *SecretRedactingEventStore) Initialize(build db.Build) error {
+    s.initializeCredVarsTrackingForBuild(build)
+    return s.EventStore.Initialize(build)
+}
+
+func (s *SecretRedactingEventStore) Put(build db.Build, events []atc.Event) error {
+    redactedEvents := make([]atc.Event, len(events)
+    for i, evt := range events {
+        redactedEvents[i] = s.redactSecrets(build, evt)
+    }
+    return s.EventStore.Put(build, redactedEvent)
+}
+```
+
+Another application is continuous syslog forwarding:
+
+```go
+type SyslogForwardingEventStore struct {
+    EventStore
+    // Some syslog forwarding state
+}
+
+func (s *SyslogForwardingEventStore) Put(build db.Build, events []atc.Event) error {
+    if err := s.syslogForwardEvents(build, events); err != nil {
+        return err
+    }
+    return s.EventStore.Put(build, event)
+}
+```
+
+How about migrating build events to cold-storage after they're completed:
```

Comment:
that's a pretty useful use case, specially for those business that require keeping logs for .. years.

assuming this would be implemented by, say, aws' s3, one can even place lifecycle policies to their bucket (https://docs.aws.amazon.com/AmazonS3/latest/dev/lifecycle-transition-general-considerations.html) and then get some sweet price reduction automatically for those super old build events